cmd/kops/get_nodes.go

```diff
@@ -0,0 +1,168 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"k8s.io/kops/pkg/validation"
+
+	"k8s.io/kops/pkg/resources/openstack"
+
+	"k8s.io/klog/v2"
+
+	"k8s.io/client-go/kubernetes"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/kops/pkg/cloudinstances"
+
+	"k8s.io/kops/pkg/resources"
+	"k8s.io/kops/util/pkg/tables"
+
+	"k8s.io/kops/pkg/apis/kops"
+	"k8s.io/kops/pkg/resources/aws"
+
+	"github.com/spf13/cobra"
+	"k8s.io/kops/cmd/kops/util"
+	"k8s.io/kops/upup/pkg/fi/cloudup"
+
+	osCloudup "k8s.io/kops/upup/pkg/fi/cloudup/openstack"
+)
+
+func NewCmdGetNodes(f *util.Factory, out io.Writer, options *GetOptions) *cobra.Command {
+
+	cmd := &cobra.Command{
+		Use:     "nodes",
+		Short:   "",
+		Long:    "",
+		Example: "",
+		Run: func(cmd *cobra.Command, args []string) {
+			ctx := context.TODO()
+
+			if err := rootCommand.ProcessArgs(args); err != nil {
+				exitWithError(err)
+			}
+
+			err := RunGetNodes(ctx, f, out, options)
+			if err != nil {
+				exitWithError(err)
+			}
+		},
+	}
+
+	return cmd
+}
+
+func RunGetNodes(ctx context.Context, f *util.Factory, out io.Writer, options *GetOptions) error {
+
+	clientset, err := f.Clientset()
+	if err != nil {
+		return err
+	}
+
+	clusterName := rootCommand.ClusterName()
+	options.clusterName = clusterName
+	if clusterName == "" {
+		return fmt.Errorf("--name is required")
+	}
+
+	cluster, err := clientset.GetCluster(ctx, options.clusterName)
+	if err != nil {
+		return err
+	}
+
+	if cluster == nil {
+		return fmt.Errorf("cluster not found %q", options.clusterName)
+	}
+
+	cloud, err := cloudup.BuildCloud(cluster)
+	if err != nil {
+		return err
+	}
+
+	contextName := cluster.ObjectMeta.Name
+	clientGetter := genericclioptions.NewConfigFlags(true)
+	clientGetter.Context = &contextName
+
+	config, err := clientGetter.ToRESTConfig()
+	if err != nil {
+		return fmt.Errorf("cannot load kubecfg settings for %q: %v", contextName, err)
+	}
+	k8sClient, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return fmt.Errorf("cannot build kubernetes api client for %q: %v", contextName, err)
+	}
+	nodeList, err := k8sClient.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		klog.V(2).Infof("error listing nodes: %v", err)
+	}
+	nodes := cloudinstances.GetNodeMap(nodeList.Items, cluster)
+
+	var instances []*resources.Instance
+	switch cloud.ProviderID() {
```

Comment:
Would it make sense to use `fi.Cloud.GetCloudGroups()` instead?

Reply:
I was thinking of using `GetCloudGroups` to output `NeedsUpdate` on instance level (also requested in #7902).
I don't think that function works without access to the API though, so it doesn't replace this switch block.

Reply:
Where does that function access the API? It's not even given an API client.

You can pass it a slice of Node, but instances don't need to be in that slice.

Reply:
This switch block is only useful if there's a bug in `GetCloudGroups` that causes it to not return instances that this block would.

Reply:
The two functions return different representation of the instance though (we may have too many different representations of an instance in kops ..). The `CloudInstanceMember` type doesn't have much of the data that is useful for this command, so we either way need to use an additional abstraction in addition.

Reply:
It might be reasonable to add the fields to `CloudInstanceMember`. The AWS ones don't look too expensive to always do for cluster validation.

The switch block is a code smell. If not `GetCloudGroups` it should probably be a new method on `fi.Cloud`.